### ISSUE TITLE
Implement feature flags support in PHP

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2316]
Site Kit already uses feature flags in JS (powered by Webpack) to enable certain features only for development builds. We should add support to achieve something similar in PHP.

Note that due to the way JS is bundled by Webpack, the JS code that is behind feature flags won't even be included in a production build. This isn't possible in PHP, but that is acceptable. We already ship a bunch of PHP code that will only be used by future features - the goal of this enhancement is to be able to easily test PHP features before enabling them in a production build.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### Infrastructure
* The `webpack.feature-flags.config.js` file should be turned into a `feature-flags.json` file (note JSON), since it's no longer purely Webpack-specific.
* Webpack should generate a JSON file `config.json` that is added within `/dist`, which simply includes the following code:  `{ "flagMode":"BUILDMODE" }` (where `BUILDMODE` is either `production` or `development`).
* There should be a `Google\Site_Kit\Core\Util\Feature_Flags` class that allows detecting whether specific feature flags are enabled.
    * The class should have a public static method `enabled( string $flag )` that returns a boolean whether the feature flag is active, by checking whether the requested feature flag's required build mode (see below) matches the current build mode (see below). Callers should be able to pass strings like `userInput` or `widgets.dashboard`.
    * The class should have a private method that reads the `feature-flags.json` file if it exists. Its content should be flattened into a map where the keys are the full `$flag` qualifiers (e.g. `widgets.dashboard`) and the values are the mode for which the feature flag is enabled.
    * The class should have a private method that returns the current flag mode. If the above `dist/config.json` file exists, it should be parsed, and the `flagMode` value should be used. If it doesn't exist or parsing fails, `production` should be assumed. That value should then be filterable via `googlesitekit_flag_mode` (which can be modified e.g. for testing).

#### Example usage
```php
if ( Feature_Flags::enabled( 'userInput' ) ) {
    // Code that only runs when feature flag is enabled.
}
```

### Initial use-case
(originally part of #2133)
* When redirecting the user to the authentication service in PHP, if the `userInput` feature flag is enabled in PHP (via the above infrastructure), the `googlesitekit_user_input_state` user option should be set to `required` if it's not already set to `completed`. Furthermore, if it's not already set to `completed`, the redirect should happen including a query parameter `mode` set to `user_input`, which will cause the copy and UI to be slightly modified to cater for the change in the flow.

### 1) Convert feature-flags js to json.
Rename `webpack.feature-flags.config.js` to `feature-flags.json`.

Convert it to JSON:
```json
{
    "featureFlags": {
        "widgets": {
            "dashboard": {
                "enabled": "development"
            }, 
            "pageDashboard": {
                "enabled": "development"
            }
        }, 
        "userInput": {
            "enabled": "development"
        }
    }
}
```

Change the existing `webpack.config.js` so that we can read our new json file:

```diff
-const flagsConfig = require( './webpack.feature-flags.config' );
+const flagsConfig = require( './feature-flags.json' );
```

### 2) Generate a config.json
Add a new dev dependency to package.json with `npm install --save-dev create-file-webpack`

Make some changes to webpack.config.js:
```js
const CreateFileWebpack = require('create-file-webpack')
```

Add a new plugin config (add it after `new ManifestPlugin`...):
```js
new CreateFileWebpack({
    path: './dist',
    fileName: 'config.json',
    content: JSON.stringify( { flagMode } ),
} ),
```

### 3) Create `Google\Site_Kit\Core\Util\Feature_Flags`
Create a file `includes/Core/Util/Feature_Flags.php`

```php
class Feature_Flags { 
    public static enabled( string $flag ) {
        $feature_flag_map = self::get_feature_flag_map();
        return isset( $feature_flag_map[ $flag ] ) ? $feature_flag_map[ $flag ] : false;
    }

    private static get_feature_flag_map() {
        // 1. Read json file feature-flags.json, return empty array if it doesn't exist. file_get_contents()
        // 2. Convert to array map. json_decode().
        // 3. We get the flag mode. $flag_mode = self::get_flag_mode();
        // 4. Starting at featureFlags, we flatten the array, With dots between nested levels, replacing any string value with whether or not it strict equals the $flag_mode.
        // Example:
        /// get_flag_mode() returns 'development'
        /// JSON File contents :
        ///{
        ///  "featureFlags": {
        ///    "foo": {
        ///      "bar": {
        ///        "baz": "development"
        ///      }
        ///    }, 
        ///    "featureX": "production"
        ///  }
        ///}
        /// Corresponds to $map[ 'foo.bar.baz' ] = true;
        /// Corresponds to $map[ 'featureX' ] = false;

    }

    private static get_flag_mode() {
        $googlesitekit_flag_mode = 'production';

        // 1. Read json file dist/config.json into array if it exists. file_get_contents() and json_decode()
        // 2. If flagMode key exists, use its value for $googlesitekit_flag_mode

        return apply_filters( 'googlesitekit_flag_mode', $googlesitekit_flag_mode );
    }


}
```

### 4) Implement logic for requiring user input

In the `Authentication` class
* Add a private method `require_user_input()` that, if the `userInput` feature flag is enabled, sets the `googlesitekit_user_input_state` user option to `required` if it's not already set to `completed`.
* Call that method in the same closure for `Google_Proxy::ACTION_SETUP` which at the end redirects the user to the proxy.

### Test Coverage

* 

### Visual Regression Changes

* 

## QA Brief

**Test `config.json` creation**
* Run `npm run build` and confirm that `dist/config.json` is json matching `{"flagMode":"production"}`
* Run `npm run dev` and confirm that `dist/config.json` is json matching `{"flagMode":"development"}`
* Run `npm run release-zip` and confirm that `dist/config.json` is **not present** in the resulting zip file

**Test setting up Site Kit with incomplete User Input**
* Using a development build, set up Site Kit after a reset
* When you get to the proxy, you should see `mode=user_input` in the URL
* After completing the proxy flow, you should land on the `googlesitekit-user-input` admin screen.

## Changelog entry

* Add support for feature flags in PHP and experimentally require user input completion upon setup.

**Comment by felixarntz** [comment]
@aaemnnosttv I've specified ACs here (fairly close to an eventual IB) based on our conversation earlier. Please review so we can get this moving.

**Comment by felixarntz** [comment]
@aaemnnosttv Can you review the ACs here so we can move this to IB soon?

**Comment by aaemnnosttv** [comment]
@felixarntz AC's look good for the most part. A few thoughts on some of the points:

> Webpack should generate a PHP file that is added somewhere within `/dist`, which simply includes the following code

If we're including a file from dist, maybe a `json` file would be better? E.g. simply `{ "flagMode":"development" }` Then that would be loaded and parsed by the feature flag class on the fly and the filter would only be relevant for tests, passing the parsed value through (with `production` as the default still). That seems a bit better (maybe safer) than loading a PHP file just to add a filter.

> `googlesitekit_build_mode`

How about `googlesitekit_flag_mode` to match the `--flag-mode` option we added to the build command?

**Comment by felixarntz** [comment]
@aaemnnosttv Sounds good. I've modified the ACs accordingly, defining that file as `dist/config.json`. It should be able to have `production` _or_ `development` as value for the `flagMode`.

**Comment by felixarntz** [comment]
IB ✅ 

@ivankruchkoff The IB looks good technically, however for the future, I'd advise to explain the approach a bit more abstractly - sometimes pseudocode can be helpful, but it shouldn't be needed for an IB to write out an entire class definition in PHP, as well as you don't need to write out things like what the JSON file looks like (that one should be fairly obvious) - it's better to just note _that_ it needs to be transformed to a JSON file.

For things like classes and methods and functions, try to summarize the intended names and how they should work on a high-level.

To be honest, for this particular issue, the ACs are very close to code already, so it's probably unavoidable to almost re-hash them in the IB anyway. But wanted to point the above things out for future IBs 👍

**Comment by aaemnnosttv** [comment]
Bumping the estimate before starting here as this seems a bit tight as-is. Confirmed the change with @eclarke1.

**Comment by aaemnnosttv** [comment]
@felixarntz please see the new issue #2452 before reviewing this as somewhat conflicts with this.

**Comment by eugene-manuilov** [comment]
QA :heavy_check_mark: all works as intended.
